datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java

```diff
@@ -0,0 +1,153 @@
+    final AtomicInteger remainder = new AtomicInteger(datastreamPartitions.getPartitions().size() % totalTaskCount);
+    LOG.debug("maxPartitionPerTask {}, task count {}", maxPartitionPerTask, totalTaskCount);
+
+    Collections.shuffle(unassignedPartitions);
+
+    Map<String, Set<DatastreamTask>> newAssignment = new HashMap<>();
+
+    //Step 2: generate new assignment. Assign unassigned partitions to tasks and create new task if there is
+    // a partition change
+    currentAssignment.keySet().forEach(instance -> {
+      Set<DatastreamTask> tasks = currentAssignment.get(instance);
+      Set<DatastreamTask> newAssignedTask = tasks.stream().map(task -> {
+        if (!dgName.equals(task.getTaskPrefix())) {
+          return task;
+        } else {
+          Set<String> newPartitions = new HashSet<>(task.getPartitionsV2());
+          newPartitions.retainAll(datastreamPartitions.getPartitions());
+
+          //We need to create new task if the partition is changed
+          boolean partitionChanged = newPartitions.size() != task.getPartitionsV2().size();
+
+          int allowedPartitions = remainder.get() > 0 ? maxPartitionPerTask + 1 : maxPartitionPerTask;
+
+          while (newPartitions.size() < allowedPartitions && unassignedPartitions.size() > 0) {
+            newPartitions.add(unassignedPartitions.remove(unassignedPartitions.size() - 1));
+            partitionChanged = true;
+          }
+
+          if (remainder.get() > 0) {
+            remainder.decrementAndGet();
+          }
+
+          if (partitionChanged) {
+            return new DatastreamTaskImpl((DatastreamTaskImpl) task, newPartitions);
+          } else {
+            return task;
+          }
+        }
+      }).collect(Collectors.toSet());
+      newAssignment.put(instance, newAssignedTask);
+    });
+    LOG.info("new assignment info, assignment: {}, all partitions: {}", newAssignment,
+        datastreamPartitions.getPartitions());
+
+    partitionSanityChecks(newAssignment, datastreamPartitions);
+    return newAssignment;
+  }
+
+  /**
+   * check if the computed assignment contains all the partitions
+   */
+  private void partitionSanityChecks(Map<String, Set<DatastreamTask>> assignedTasks,
+      DatastreamGroupPartitionsMetadata allPartitions) {
+    int total = 0;
+
+    List<String> unassignedPartitions = new ArrayList<>(allPartitions.getPartitions());
+    String datastreamGroupName = allPartitions.getDatastreamGroup().getName();
+    for (Set<DatastreamTask> tasksSet : assignedTasks.values()) {
+      for (DatastreamTask task : tasksSet) {
+        if (datastreamGroupName.equals(task.getTaskPrefix())) {
+          total += task.getPartitionsV2().size();
+          unassignedPartitions.removeAll(task.getPartitionsV2());
+        }
+      }
+    }
+    if (total != allPartitions.getPartitions().size()) {
+      throw new DatastreamRuntimeException(String.format("Validation failed after assignment, assigned partitions "
```

Comment:
typically we log before we throw the exception. It's possible that it might return in duplicate logs.  it just is a defensive pattern that we have been following, because we have seen instances where the exception thrown result in process crash and the exception is not logged.

Reply:
Has fixed the logging.